src/contexts/Application.js

```diff
@@ -0,0 +1,54 @@
+  
+import React, { createContext, useContext, useReducer, useMemo, useCallback, useEffect } from "react";
+
+const TOGGLE_WALLET_CONNECTION_MODAL = "TOGGLE_WALLET_CONNECTION_MODAL";
+const WALLET_MODAL_OPEN = "WALLET_CONNECTION_MODAL_OPEN";
+
+
+const ApplicationContext = createContext();
+
+function useApplicationContext() {
+    return useContext(ApplicationContext);
+}
+
+function reducer(state, { type, payload }) {
+    switch (type) {
+      case TOGGLE_WALLET_CONNECTION_MODAL: {
+        return { ...state, [WALLET_MODAL_OPEN]: !state[WALLET_MODAL_OPEN] }
+      }
+      default: {
+        throw Error(`Unexpected action type in ApplicationContext reducer: '${type}'.`)
+      }
+    }
+}
+
+export default function Provider({ children }) {
+    const [state, dispatch] = useReducer(reducer, {
+      [WALLET_MODAL_OPEN]: false
+    })
+  
+    const toggleWalletConnectionModal = useCallback(() => {
+      dispatch({ type: TOGGLE_WALLET_CONNECTION_MODAL })
+    }, [])
+  
+    return (
+      <ApplicationContext.Provider
+        value={useMemo(() => [state, { toggleWalletConnectionModal }], [
```

Comment:
Does `useMemo` memoize data? Just courious

Reply:
Yes it does. In this case I would think it memoizes the default context value I think